File: scripts/lib/atproto/seed.py
"""Seeding helpers for Garazyk ATProto demo and scenario scripts.

These functions encode the conventions shared by demo seeders: UTC timestamp
formatting, server readiness polling, account creation with login fallback,
idempotent record creation, chat conversation helpers, and admin-token lookup.
They are intentionally thin wrappers over XrpcClient so seed scripts remain
easy to compare against the lexicon methods they exercise.
"""

from __future__ import annotations

import time
from datetime import datetime, timezone
from typing import Any, Optional

import requests

from .transport import XrpcError
from .client import XrpcClient


# ── Timestamp ───────────────────────────────────────────────────────────────

def now_iso() -> str:
    """Return the current UTC time in the ATProto timestamp format."""
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


# ── Server health ──────────────────────────────────────────────────────────

def wait_for_server(base_url: str, timeout: int = 30) -> None:
    """Wait for a PDS to be healthy by polling /_health.

    Raises RuntimeError if the server is not healthy within the timeout.
    """
    deadline = time.time() + timeout
    last_error: str | None = None
    while time.time() < deadline:
        try:
            r = requests.get(f"{base_url}/_health", timeout=2)
            if r.status_code == 200:
                return
            last_error = f"HTTP {r.status_code}"
        except requests.RequestException as e:
            last_error = str(e)
        time.sleep(0.5)
    raise RuntimeError(f"PDS not ready at {base_url} (last: {last_error})")


def wait_for_http(url: str, timeout: int = 30, label: str = "") -> bool:
    """Wait for an HTTP endpoint to return a non-5xx response.

    Returns True if the endpoint is reachable before timeout, False otherwise.
    This looser condition is useful for endpoints that require authentication
    but still prove the service is listening once they stop returning 5xx.
    """
    label = label or url
    deadline = time.time() + timeout
    while time.time() < deadline:
        try:
            r = requests.get(url, timeout=2)
            if r.status_code < 500:
                return True
        except requests.ConnectionError:
            pass
        time.sleep(0.5)
    return False


# ── Account creation with login fallback ────────────────────────────────────

def create_account_or_login(
    client: XrpcClient,
    handle: str,
    email: str,
    password: str,
) -> dict[str, Any]:
    """Create an account, or fall back to login if it already exists.

    Returns a session dict with did, handle, accessJwt, and refreshJwt. The
    fallback keeps local seeding repeatable when test data from a previous run
    has not been removed.
    """
    try:
        return client.accounts.create_account(handle, email, password)
    except XrpcError:
        # Account might already exist -- try login
        pass
    return client.accounts.create_session(handle, password)


# ── Record creation with idempotency ───────────────────────────────────────

def create_record_idempotent(
    client: XrpcClient,
    repo: str,
    collection: str,
    record: dict[str, Any],
    token: str,
) -> dict[str, Any]:
    """Create a record, ignoring "already exists" errors.

    Returns the createRecord response on success, or an empty dict when the
    server reports an idempotent duplicate. Other XRPC failures still propagate
    so callers notice schema, auth, and storage errors.
    """
    try:
        return client.create_record(repo, collection, record, token)
    except XrpcError as exc:
        if exc.status == 400 and "already exists" in str(exc.body).lower():
            return {}
        raise


# ── Chat helpers ────────────────────────────────────────────────────────────

def get_convo_for_members(
    client: XrpcClient,
    jwt: str,
    member_dids: list[str],
) -> dict[str, Any]:
    """Get or create a conversation for the given member DIDs.

    Returns the raw response dict, which usually contains ``convo.id``. The
    chat service creates the conversation if an exact membership match does not
    already exist.
    """
    return client.xrpc_post(
        "chat.bsky.convo.getConvoForMembers",
        {"members": member_dids},
        token=jwt,
    )


def send_message(
    client: XrpcClient,
    jwt: str,
    convo_id: str,
    text: str,
) -> dict[str, Any]:
    """Send a text message in a conversation.

    Returns the sendMessage response. The helper stamps messages at send time
    so scenario scripts can describe conversation content without repeating
    timestamp boilerplate.
    """
    return client.xrpc_post(
        "chat.bsky.convo.sendMessage",
        {
            "convoId": convo_id,
            "message": {
                "$type": "chat.bsky.convo.def#messageRef",
                "text": text,
                "createdAt": now_iso(),
            },
        },
        token=jwt,
    )


def list_convos(
    client: XrpcClient,
    jwt: str,
    limit: int = 20,
) -> dict[str, Any]:
    """List conversations for the authenticated user."""
    return client.xrpc_get(
        "chat.bsky.convo.listConvos",
        {"limit": limit},
        token=jwt,
    )


def get_messages(
    client: XrpcClient,
    jwt: str,
    convo_id: str,
    limit: int = 50,
) -> dict[str, Any]:
    """Get messages for a conversation."""
    return client.xrpc_get(
        "chat.bsky.convo.getMessages",
        {"convoId": convo_id, "limit": limit},
        token=jwt,
    )


# ── Admin helpers ───────────────────────────────────────────────────────────

def get_pds_admin_token(base_url: str, password: str) -> str:
    """Get a PDS admin JWT via POST /admin/login.

    Returns the bearer token string. Raises RuntimeError instead of XrpcError
    because /admin/login is an admin HTTP route rather than an XRPC method.
    """
    r = requests.post(
        f"{base_url}/admin/login",
        json={"password": password},
        timeout=10,
    )
    if r.status_code != 200:
        raise RuntimeError(f"PDS admin login failed: {r.status_code}")
    return r.json()["token"]

File: packages/gruszka/seed.ts
/** Seed data helpers for ATProto E2E tests — accounts, posts, chat interactions. @module seed */
import { XrpcClient } from "./client.ts";
import { XrpcError } from "./transport.ts";
import type { GeneratedClient as ExactGeneratedClient } from "./lexicons.ts";
import type { ProcedureOutput, QueryOutput } from "./lexicons.ts";

/** Default test accounts used by seed data helpers. */
export const DEFAULT_ACCOUNTS = [
  { handle: "alice.test", email: "alice@test.local", password: "alicepass" },
  { handle: "bob.test", email: "bob@test.local", password: "bobpass" },
  { handle: "carol.test", email: "carol@test.local", password: "carolpass" },
];

/** Post text templates for seed data. Use `{handle}` as a placeholder for the account handle. */
export const DEFAULT_POSTS_TEMPLATES = [
  "Hello from {handle}! Excited to be on the ATProto network!",
  "Just set up my PDS instance. Decentralization rocks!",
  "Working on some cool features today. #atproto #coding",
  "Beautiful day to build something new!",
  "The future of social is decentralized. Here we go!",
  "Just learned about MST (Merkle Search Tree) -- fascinating tech!",
  "Shoutout to the Bluesky team for the protocol design!",
  "Testing out the firehose relay functionality today.",
  "Record indexing is working great with the new backfill logic.",
  "Admin UI makes managing the PDS so much easier!",
];

/** Current timestamp in ISO 8601 format with milliseconds stripped (e.g. `2024-01-15T12:30:00Z`). */
export function nowIso(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function exactApi(client: XrpcClient): ExactGeneratedClient {
  return client.api as unknown as ExactGeneratedClient;
}

/**
 * Wait for a PDS server to respond with HTTP 200 on its `/_health` endpoint.
 *
 * @param baseUrl - Root URL of the PDS (e.g. `http://localhost:2583`)
 * @param timeout - Maximum seconds to wait. @defaultValue 30
 * @throws If the server does not become healthy within the timeout
 */
export async function waitForServer(
  baseUrl: string,
  timeout = 30,
): Promise<void> {
  const deadline = Date.now() + timeout * 1000;
  let lastError = "not attempted";
  while (Date.now() < deadline) {
    try {
      const resp = await fetch(`${baseUrl.replace(/\/$/, "")}/_health`);
      if (resp.status === 200) return;
      lastError = `HTTP ${resp.status}`;
    } catch (exc) {
      lastError = exc instanceof Error ? exc.message : String(exc);
    }
    await new Promise((resolve) => setTimeout(resolve, 500));
  }
  throw new Error(`PDS not ready at ${baseUrl} (last: ${lastError})`);
}

/**
 * Create an account, or fall back to logging in if the handle already exists.
 *
 * @param client - XRPC client pointing at the PDS
 * @param handle - Desired handle (e.g. `alice.test`)
 * @param email - Account email
 * @param password - Account password
 * @throws {XrpcError} If both creation and login fail.
 */
export async function createAccountOrLogin(
  client: XrpcClient,
  handle: string,
  email: string,
  password: string,
): Promise<ProcedureOutput<"com.atproto.server.createSession">> {
  const api = exactApi(client);
  try {
    return await api.com.atproto.server.createAccount({
      email,
      handle,
      password,
    }) as unknown as ProcedureOutput<"com.atproto.server.createSession">;
  } catch (exc) {
    if (exc instanceof XrpcError && exc.status === 400) {
      const body = typeof exc.body === "string"
        ? exc.body
        : JSON.stringify(exc.body);
      if (body.toLowerCase().includes("already exists")) {
        return await api.com.atproto.server.createSession({
          identifier: handle,
          password,
        });
      }
    }
    throw exc;
  }
}

/**
 * Create a repository record, returning an empty object if it already exists.
 *
 * Catches `XrpcError` with status 400 containing "already exists" and returns `{}`.
 *
 * @param client - XRPC client pointing at the PDS
 * @param repo - DID of the repository
 * @param collection - NSID collection name (e.g. `app.bsky.feed.post`)
 * @param record - Record body to create
 * @param token - Auth access JWT
 * @throws {XrpcError} If the creation fails with an error other than "already exists"
 */
export async function createRecordIdempotent(
  client: XrpcClient,
  repo: string,
  collection: string,
  record: Record<string, unknown>,
  token: string,
): Promise<
  ProcedureOutput<"com.atproto.repo.createRecord"> | Record<string, never>
> {
  const api = exactApi(client);
  try {
    return await api.com.atproto.repo.createRecord({
      repo,
      collection,
      record,
    }, token);
  } catch (exc) {
    if (exc instanceof XrpcError && exc.status === 400) {
      const body = typeof exc.body === "string"
        ? exc.body
        : JSON.stringify(exc.body);
      if (body.toLowerCase().includes("already exists")) return {};
    }
    throw exc;
  }
}

/** Context for interacting with the Bluesky chat service via XRPC. */
export interface ChatServiceContext {
  /** XRPC client for the PDS (used for service auth tokens). */
  pdsClient: XrpcClient;
  /** XRPC client for the chat service endpoint. */
  chatClient: XrpcClient;
  /** DID of the chat service (e.g. `did:web:localhost#bsky_chat`). */
  serviceDid: string;
}

/**
 * Derive the chat service DID from a URL or an explicit override.
 *
 * If `configured` is provided, uses it directly (appending `#bsky_chat` if no
 * fragment is present). Otherwise, constructs a `did:web` DID from the URL
 * hostname, mapping `127.0.0.1`/`::1` to `localhost`.
 *
 * @param baseUrl - Chat service URL
 * @param configured - Optional explicit DID override
 * @returns The chat service DID
 */
export function chatServiceDidForUrl(
  baseUrl: string,
  configured?: string,
): string {
  if (configured?.trim()) {
    const serviceDid = configured.trim();
    return serviceDid.includes("#") ? serviceDid : `${serviceDid}#bsky_chat`;
  }

  const url = new URL(baseUrl);
  const hostname = url.hostname === "127.0.0.1" || url.hostname === "::1"
    ? "localhost"
    : url.hostname;
  const isDefaultPort = !url.port ||
    (url.protocol === "https:" && url.port === "443") ||
    (url.protocol === "http:" && url.port === "80");
  const didHost = isDefaultPort ? hostname : `${hostname}%3A${url.port}`;
  return `did:web:${didHost}#bsky_chat`;
}

/**
 * Create a {@link ChatServiceContext} with a chat client and derived service DID.
 *
 * @param pdsClient - XRPC client for the PDS
 * @param chatUrl - Chat service base URL
 * @param configuredServiceDid - Optional explicit DID override
 * @returns A fully initialized chat service context
 */
export function createChatServiceContext(
  pdsClient: XrpcClient,
  chatUrl: string,
  configuredServiceDid?: string,
): ChatServiceContext {
  return {
    pdsClient,
    chatClient: new XrpcClient(chatUrl),
    serviceDid: chatServiceDidForUrl(chatUrl, configuredServiceDid),
  };
}

/**
 * Obtain a scoped auth token for a specific chat XRPC method.
 *
 * Calls `com.atproto.server.getServiceAuth` on the PDS to get a token scoped
 * to the chat service DID and the requested method.
 *
 * @param context - Chat service context
 * @param accessJwt - PDS access JWT for the calling user
 * @param method - XRPC method NSID (e.g. `chat.bsky.convo.sendMessage`)
 * @returns A scoped auth token for the chat service
 * @throws If the PDS does not return a token
 */
export async function chatServiceAuthForMethod(
  context: ChatServiceContext,
  accessJwt: string,
  method: string,
): Promise<string> {
  const response = await context.pdsClient.api.com.atproto.server
    .getServiceAuth({
      aud: context.serviceDid,
      lxm: method,
    }, accessJwt);
  const token = String((response as any)?.token || "");
  if (!token) {
    throw new Error(
      `com.atproto.server.getServiceAuth did not return a token for ${method}`,
    );
  }
  return token;
}

/**
 * Perform an authenticated chat XRPC GET with automatic service auth.
 *
 * @param context - Chat service context
 * @param accessJwt - PDS access JWT for the calling user
 * @param method - XRPC method NSID
 * @param params - Query parameters
 */
export async function chatXrpcGet(
  context: ChatServiceContext,
  accessJwt: string,
  method: string,
  params?: Record<string, unknown>,
): Promise<any> {
  const token = await chatServiceAuthForMethod(context, accessJwt, method);
  return await context.chatClient.raw.xrpcGet(method, params, token);
}

/**
 * Perform an authenticated chat XRPC POST with automatic service auth.
 *
 * @param context - Chat service context
 * @param accessJwt - PDS access JWT for the calling user
 * @param method - XRPC method NSID
 * @param body - Request body
 */
export async function chatXrpcPost(
  context: ChatServiceContext,
  accessJwt: string,
  method: string,
  body?: Record<string, unknown>,
): Promise<any> {
  const token = await chatServiceAuthForMethod(context, accessJwt, method);
  return await context.chatClient.raw.xrpcPost(method, body, token);
}

/**
 * Get or create a DM conversation for a set of members.
 *
 * @param context - Chat service context
 * @param accessJwt - PDS access JWT for the calling user
 * @param memberDids - DIDs of the conversation members
 */
export async function chatGetConvoForMembers(
  context: ChatServiceContext,
  accessJwt: string,
  memberDids: string[],
): Promise<QueryOutput<"chat.bsky.convo.getConvoForMembers">> {
  return await chatXrpcGet(
    context,
    accessJwt,
    "chat.bsky.convo.getConvoForMembers",
    {
      members: memberDids,
    },
  );
}

/**
 * Send a text message in a chat conversation.
 *
 * @param context - Chat service context
 * @param accessJwt - PDS access JWT for the calling user
 * @param convoId - Conversation ID
 * @param text - Message text
 */
export async function chatSendMessage(
  context: ChatServiceContext,
  accessJwt: string,
  convoId: string,
  text: string,
): Promise<ProcedureOutput<"chat.bsky.convo.sendMessage">> {
  return await chatXrpcPost(context, accessJwt, "chat.bsky.convo.sendMessage", {
    convoId,
    message: { text },
  });
}

/**
 * List conversations for the authenticated user.
 *
 * @param context - Chat service context
 * @param accessJwt - PDS access JWT for the calling user
 * @param limit - Maximum number of conversations to return. @defaultValue 20
 */
export async function chatListConvos(
  context: ChatServiceContext,
  accessJwt: string,
  limit = 20,
): Promise<QueryOutput<"chat.bsky.convo.listConvos">> {
  return await chatXrpcGet(context, accessJwt, "chat.bsky.convo.listConvos", {
    limit,
  });
}

/**
 * Get messages in a chat conversation.
 *
 * @param context - Chat service context
 * @param accessJwt - PDS access JWT for the calling user
 * @param convoId - Conversation ID
 * @param limit - Maximum number of messages to return. @defaultValue 50
 */
export async function chatGetMessages(
  context: ChatServiceContext,
  accessJwt: string,
  convoId: string,
  limit = 50,
): Promise<QueryOutput<"chat.bsky.convo.getMessages">> {
  return await chatXrpcGet(context, accessJwt, "chat.bsky.convo.getMessages", {
    convoId,
    limit,
  });
}

/** Get or create a DM conversation for a set of members (direct XRPC, no service auth). */
export async function getConvoForMembers(
  client: XrpcClient,
  jwt: string,
  memberDids: string[],
): Promise<QueryOutput<"chat.bsky.convo.getConvoForMembers">> {
  return await exactApi(client).chat.bsky.convo.getConvoForMembers({
    members: memberDids,
  }, jwt);
}

/** Send a text message in a chat conversation (direct XRPC, no service auth). */
export async function sendMessage(
  client: XrpcClient,
  jwt: string,
  convoId: string,
  text: string,
): Promise<ProcedureOutput<"chat.bsky.convo.sendMessage">> {
  return await exactApi(client).chat.bsky.convo.sendMessage({
    convoId,
    message: { text },
  }, jwt);
}

/** List conversations for the authenticated user (direct XRPC, no service auth). */
export async function listConvos(
  client: XrpcClient,
  jwt: string,
  limit = 20,
): Promise<QueryOutput<"chat.bsky.convo.listConvos">> {
  return await exactApi(client).chat.bsky.convo.listConvos({ limit }, jwt);
}

/** Get messages in a chat conversation (direct XRPC, no service auth). */
export async function getMessages(
  client: XrpcClient,
  jwt: string,
  convoId: string,
  limit = 50,
): Promise<QueryOutput<"chat.bsky.convo.getMessages">> {
  return await exactApi(client).chat.bsky.convo.getMessages(
    { convoId, limit },
    jwt,
  );
}
